File: src/app/wasm-hello-window/hello.r
/*
 * hello.r — minimal Rez resources for the Phase B mixed-build demo
 * (cv-mac #100 Phase B).
 *
 * Two resources:
 *   - WIND 128 — the window hello.c loads via GetNewWindow(128, ...)
 *   - SIZE -1 — override the 1 MB default heap (we don't need that
 *               much; 256 KB is plenty for an empty window).
 *
 * This is deliberately the smallest .r that can demonstrate the
 * mixed-build path. Hello Mac's .r has the full Finder-binding ritual
 * (BNDL, FREF, ICN#, vers, MBAR, MENUs, ALRT/DITL); we skip all of
 * that — the demo is "see a window with text, click to dismiss."
 *
 * The Resource Manager merges this fork onto the C-built MacBinary's
 * fork via spliceResourceFork — the user's resources win on (type, id)
 * collisions, so our SIZE -1 replaces the default makeRetro68DefaultSizeFork().
 */

#include "Processes.r"
#include "Windows.r"
#include "MacTypes.r"

/* Signature resource. Required by the splice path's Type/Creator
 * lock-check (runBuild rejects builds where the signature line is
 * missing). 'CVWW' = Classic Vibe Wasm Window. */
data 'CVWW' (0, "Owner signature") {
    "CVWW"
};

/* Window definition. 60,60 → 240,380 = 180 tall, 320 wide, comfortable
 * room for a single line of centred text. documentProc gives the
 * standard System 7 title bar + close box; goAway makes the close box
 * functional; visible means the window shows immediately on
 * GetNewWindow (we still call ShowWindow for explicitness). */
resource 'WIND' (128) {
    { 60, 60, 240, 380 },
    documentProc,
    visible,
    goAway,
    0,
    "Wasm Hello — Phase B",
    noAutoCenter
};

/* SIZE resource (-1) — overrides the default 1 MB heap. 256 KB
 * preferred + minimum is more than enough for a Toolbox-only app
 * that draws one string. Flag 0x0080 = "32-bit clean" (System 7+
 * required this for apps that use addresses > 24 bits, which any
 * modern app should). */
data 'SIZE' (-1, "Wasm Hello Window") {
    $"0080"                /* flags: 32-bit clean */
    $"00040000"            /* preferred: 256 KB */
    $"00040000"            /* minimum: 256 KB */
};

File: src/app/wasm-hello-window/hello.c
/*
 * hello.c — windowed Hello, mixed C + .r demo (cv-mac #100 Phase B).
 *
 * Pairs with hello.r. Demonstrates the in-browser path where BOTH the
 * .c source AND the .r resources are compiled client-side and spliced
 * into one MacBinary. The resource fork's WIND, MBAR, MENU, etc.
 * resources come from hello.r; the data fork's CODE 0..N + RELA + a
 * default SIZE come from cc1+as+ld+Elf2Mac; spliceResourceFork merges
 * them (user-wins on collision, so this app's SIZE overrides the
 * default 1 MB heap from runBuildInBrowserC).
 *
 * Differs from wasm-hello/hello.c (no-window demo) by having an actual
 * window resource — gets a proper title bar, content area, the works.
 * The window's ID is 128 (matching the WIND resource declared in
 * hello.r). NewWindow(...) loads it from the resource fork at runtime.
 */

#include <Types.h>
#include <Quickdraw.h>
#include <Fonts.h>
#include <Windows.h>
#include <Menus.h>
#include <TextEdit.h>
#include <Dialogs.h>
#include <Events.h>
#include <Memory.h>

#define kWindowID 128

QDGlobals qd;

static const unsigned char HELLO_PSTR[] = {
    13, 'H', 'e', 'l', 'l', 'o', ',', ' ',
    'W', 'o', 'r', 'l', 'd', '!',
};

int main(void) {
    /* Standard Toolbox init. */
    InitGraf(&qd.thePort);
    InitFonts();
    InitWindows();
    InitMenus();
    TEInit();
    InitDialogs(0);

    /* GetNewWindow loads our window definition from the resource fork
     * (WIND 128 declared in hello.r). The (-1) puts it on top of the
     * window list; (WindowPtr)NULL passes no behind-window placement. */
    WindowPtr win = GetNewWindow(kWindowID, NULL, (WindowPtr)(-1));
    if (!win) {
        SysBeep(10);
        return 1;
    }
    SetPort((GrafPtr)win);
    ShowWindow(win);

    /* Draw "Hello, World!" centred-ish in the window. */
    MoveTo(120, 80);
    DrawString(HELLO_PSTR);

    /* Event loop: redraw on update events, quit on click. */
    for (;;) {
        EventRecord ev;
        WaitNextEvent(everyEvent, &ev, 60, NULL);
        switch (ev.what) {
            case mouseDown:
                if (FindWindow(ev.where, &win) == inGoAway) {
                    return 0;  /* click close box → exit */
                }
                return 0;  /* any other click exits too */
            case updateEvt:
                if ((WindowPtr)ev.message == win) {
                    BeginUpdate(win);
                    MoveTo(120, 80);
                    DrawString(HELLO_PSTR);
                    EndUpdate(win);
                }
                break;
        }
    }
    return 0;
}
